File: uis/client/mobile_app/lib/models/trade_finance.dart
class LetterOfCredit {
  final String id;
  final String lcRef;
  final String type; // irrevocable, revocable, standby, confirmed
  final String applicantId;
  final String beneficiaryName;
  final String beneficiaryCountry;
  final double amount;
  final String currency;
  final String availableBy;
  final String expiryDate;
  final String placeOfExpiry;
  final String goodsDescription;
  final String swiftRef;
  final String status; // draft, issued, confirmed, amended, expired, cancelled
  final String? issuedAt;
  final String createdAt;

  LetterOfCredit({
    this.id = '',
    this.lcRef = '',
    this.type = 'irrevocable',
    this.applicantId = '',
    this.beneficiaryName = '',
    this.beneficiaryCountry = '',
    this.amount = 0.0,
    this.currency = 'USD',
    this.availableBy = '',
    this.expiryDate = '',
    this.placeOfExpiry = '',
    this.goodsDescription = '',
    this.swiftRef = '',
    this.status = 'draft',
    this.issuedAt,
    this.createdAt = '',
  });

  factory LetterOfCredit.fromJson(Map<String, dynamic> json) {
    return LetterOfCredit(
      id: json['id']?.toString() ?? '',
      lcRef: json['lc_ref']?.toString() ?? json['lcRef']?.toString() ?? '',
      type: json['type']?.toString() ?? 'irrevocable',
      applicantId: json['applicant_id']?.toString() ?? json['applicantId']?.toString() ?? '',
      beneficiaryName: json['beneficiary_name']?.toString() ?? json['beneficiaryName']?.toString() ?? '',
      beneficiaryCountry: json['beneficiary_country']?.toString() ?? json['beneficiaryCountry']?.toString() ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      currency: json['currency']?.toString() ?? 'USD',
      availableBy: json['available_by']?.toString() ?? json['availableBy']?.toString() ?? '',
      expiryDate: json['expiry_date']?.toString() ?? json['expiryDate']?.toString() ?? '',
      placeOfExpiry: json['place_of_expiry']?.toString() ?? json['placeOfExpiry']?.toString() ?? '',
      goodsDescription: json['goods_description']?.toString() ?? json['goodsDescription']?.toString() ?? '',
      swiftRef: json['swift_ref']?.toString() ?? json['swiftRef']?.toString() ?? '',
      status: json['status']?.toString() ?? 'draft',
      issuedAt: json['issued_at']?.toString() ?? json['issuedAt']?.toString(),
      createdAt: json['created_at']?.toString() ?? json['createdAt']?.toString() ?? '',
    );
  }
}

class BankGuarantee {
  final String id;
  final String guaranteeRef;
  final String type; // performance, payment, bid_bond, advance_payment
  final String applicantId;
  final String beneficiaryName;
  final double amount;
  final String currency;
  final String expiryDate;
  final String purpose;
  final String status; // draft, issued, extended, cancelled, expired
  final String createdAt;

  BankGuarantee({
    this.id = '',
    this.guaranteeRef = '',
    this.type = 'performance',
    this.applicantId = '',
    this.beneficiaryName = '',
    this.amount = 0.0,
    this.currency = 'NGN',
    this.expiryDate = '',
    this.purpose = '',
    this.status = 'draft',
    this.createdAt = '',
  });

  factory BankGuarantee.fromJson(Map<String, dynamic> json) {
    return BankGuarantee(
      id: json['id']?.toString() ?? '',
      guaranteeRef: json['guarantee_ref']?.toString() ?? json['guaranteeRef']?.toString() ?? '',
      type: json['type']?.toString() ?? 'performance',
      applicantId: json['applicant_id']?.toString() ?? json['applicantId']?.toString() ?? '',
      beneficiaryName: json['beneficiary_name']?.toString() ?? json['beneficiaryName']?.toString() ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      currency: json['currency']?.toString() ?? 'NGN',
      expiryDate: json['expiry_date']?.toString() ?? json['expiryDate']?.toString() ?? '',
      purpose: json['purpose']?.toString() ?? '',
      status: json['status']?.toString() ?? 'draft',
      createdAt: json['created_at']?.toString() ?? json['createdAt']?.toString() ?? '',
    );
  }
}

class FactoringApplication {
  final String id;
  final String applicationRef;
  final String applicantId;
  final String debtorName;
  final double invoiceTotal;
  final double factoringAmount;
  final String currency;
  final double discountRate;
  final String status; // pending, approved, rejected, disbursed
  final int invoiceCount;
  final String createdAt;

  FactoringApplication({
    this.id = '',
    this.applicationRef = '',
    this.applicantId = '',
    this.debtorName = '',
    this.invoiceTotal = 0.0,
    this.factoringAmount = 0.0,
    this.currency = 'NGN',
    this.discountRate = 0.0,
    this.status = 'pending',
    this.invoiceCount = 0,
    this.createdAt = '',
  });

  factory FactoringApplication.fromJson(Map<String, dynamic> json) {
    return FactoringApplication(
      id: json['id']?.toString() ?? '',
      applicationRef: json['application_ref']?.toString() ?? json['applicationRef']?.toString() ?? '',
      applicantId: json['applicant_id']?.toString() ?? json['applicantId']?.toString() ?? '',
      debtorName: json['debtor_name']?.toString() ?? json['debtorName']?.toString() ?? '',
      invoiceTotal: (json['invoice_total'] as num?)?.toDouble() ?? 0.0,
      factoringAmount: (json['factoring_amount'] as num?)?.toDouble() ?? 0.0,
      currency: json['currency']?.toString() ?? 'NGN',
      discountRate: (json['discount_rate'] as num?)?.toDouble() ?? 0.0,
      status: json['status']?.toString() ?? 'pending',
      invoiceCount: (json['invoice_count'] as num?)?.toInt() ?? 0,
      createdAt: json['created_at']?.toString() ?? json['createdAt']?.toString() ?? '',
    );
  }
}

File: uis/client/web2/src/models/trade_finance.ts
export interface LetterOfCredit {
  id: string;
  lcRef: string;
  type: string; // irrevocable | revocable | standby | confirmed
  applicantId: string;
  beneficiaryName: string;
  beneficiaryCountry: string;
  amount: number;
  currency: string;
  availableBy: string;
  expiryDate: string;
  placeOfExpiry: string;
  goodsDescription: string;
  swiftRef: string;
  status: string; // draft | issued | confirmed | amended | expired | cancelled
  issuedAt?: string;
  createdAt: string;
}

export interface BankGuarantee {
  id: string;
  guaranteeRef: string;
  type: string; // performance | payment | bid_bond | advance_payment
  applicantId: string;
  beneficiaryName: string;
  amount: number;
  currency: string;
  expiryDate: string;
  purpose: string;
  status: string; // draft | issued | extended | cancelled | expired
  createdAt: string;
}

export interface FactoringApplication {
  id: string;
  applicationRef: string;
  applicantId: string;
  debtorName: string;
  invoiceTotal: number;
  factoringAmount: number;
  currency: string;
  discountRate: number;
  status: string; // pending | approved | rejected | disbursed
  invoiceCount: number;
  createdAt: string;
}

export function lcFromJson(json: Record<string, any>): LetterOfCredit {
  return {
    id: String(json.id ?? ''),
    lcRef: String(json.lc_ref ?? json.lcRef ?? ''),
    type: String(json.type ?? 'irrevocable'),
    applicantId: String(json.applicant_id ?? json.applicantId ?? ''),
    beneficiaryName: String(json.beneficiary_name ?? json.beneficiaryName ?? ''),
    beneficiaryCountry: String(json.beneficiary_country ?? json.beneficiaryCountry ?? ''),
    amount: Number(json.amount ?? 0),
    currency: String(json.currency ?? 'USD'),
    availableBy: String(json.available_by ?? json.availableBy ?? ''),
    expiryDate: String(json.expiry_date ?? json.expiryDate ?? ''),
    placeOfExpiry: String(json.place_of_expiry ?? json.placeOfExpiry ?? ''),
    goodsDescription: String(json.goods_description ?? json.goodsDescription ?? ''),
    swiftRef: String(json.swift_ref ?? json.swiftRef ?? ''),
    status: String(json.status ?? 'draft'),
    issuedAt: json.issued_at ? String(json.issued_at) : json.issuedAt ? String(json.issuedAt) : undefined,
    createdAt: String(json.created_at ?? json.createdAt ?? ''),
  };
}

export function bgFromJson(json: Record<string, any>): BankGuarantee {
  return {
    id: String(json.id ?? ''),
    guaranteeRef: String(json.guarantee_ref ?? json.guaranteeRef ?? ''),
    type: String(json.type ?? 'performance'),
    applicantId: String(json.applicant_id ?? json.applicantId ?? ''),
    beneficiaryName: String(json.beneficiary_name ?? json.beneficiaryName ?? ''),
    amount: Number(json.amount ?? 0),
    currency: String(json.currency ?? 'NGN'),
    expiryDate: String(json.expiry_date ?? json.expiryDate ?? ''),
    purpose: String(json.purpose ?? ''),
    status: String(json.status ?? 'draft'),
    createdAt: String(json.created_at ?? json.createdAt ?? ''),
  };
}

export function factoringFromJson(json: Record<string, any>): FactoringApplication {
  return {
    id: String(json.id ?? ''),
    applicationRef: String(json.application_ref ?? json.applicationRef ?? ''),
    applicantId: String(json.applicant_id ?? json.applicantId ?? ''),
    debtorName: String(json.debtor_name ?? json.debtorName ?? ''),
    invoiceTotal: Number(json.invoice_total ?? json.invoiceTotal ?? 0),
    factoringAmount: Number(json.factoring_amount ?? json.factoringAmount ?? 0),
    currency: String(json.currency ?? 'NGN'),
    discountRate: Number(json.discount_rate ?? json.discountRate ?? 0),
    status: String(json.status ?? 'pending'),
    invoiceCount: Number(json.invoice_count ?? json.invoiceCount ?? 0),
    createdAt: String(json.created_at ?? json.createdAt ?? ''),
  };
}
